File: src/services/api/providers/openai.ts
import OpenAI from "openai";
import type { ApiKeyStore } from "src/secrets/apiKeyStore";
import { resolveStoredApiKey } from "src/secrets/apiKeyStore";
import { toOpenAIFunction } from "../toolSchema";
import type { CallModelParams, Provider, ProviderEvent } from "../types";

export class OpenAIProvider implements Provider {
  readonly name = "openai";
  constructor(
    private env: NodeJS.ProcessEnv,
    private apiKeyStore?: ApiKeyStore,
  ) {}

  async listModels(): Promise<string[]> {
    const apiKey = await resolveStoredApiKey("openai", this.env, this.apiKeyStore);
    if (!apiKey) throw new Error("OPENAI_API_KEY not set");
    const client = new OpenAI({
      apiKey,
      ...(this.env.OPENAI_BASE_URL ? { baseURL: this.env.OPENAI_BASE_URL } : {}),
    });
    const res = await client.models.list();
    return res.data
      .map((m) => m.id)
      .filter((id): id is string => typeof id === "string" && id.length > 0)
      .sort((a, b) => a.localeCompare(b));
  }

  async *stream(params: CallModelParams): AsyncGenerator<ProviderEvent> {
    const apiKey = await resolveStoredApiKey("openai", this.env, this.apiKeyStore);
    if (!apiKey) {
      yield { type: "error", error: new Error("OPENAI_API_KEY not set") };
      return;
    }
    const client = new OpenAI({
      apiKey,
      ...(this.env.OPENAI_BASE_URL ? { baseURL: this.env.OPENAI_BASE_URL } : {}),
    });
    const model = params.model ?? this.env.OPENAI_MODEL ?? "gpt-4o-mini";
    const toolDefs = params.tools.map((t) => toOpenAIFunction(t));

    const messages: OpenAI.ChatCompletionMessageParam[] = [
      { role: "system", content: params.system },
      ...params.messages.map((m): OpenAI.ChatCompletionMessageParam => {
        if (m.role === "user") return { role: "user", content: m.content };
        if (m.role === "assistant") {
          const text = m.content
            .filter((b): b is { type: "text"; text: string } => b.type === "text")
            .map((b) => b.text)
            .join("");
          const toolCalls = m.content
            .filter(
              (b): b is { type: "tool_use"; id: string; name: string; input: unknown } =>
                b.type === "tool_use",
            )
            .map((b) => ({
              id: b.id,
              type: "function" as const,
              function: { name: b.name, arguments: JSON.stringify(b.input) },
            }));
          return {
            role: "assistant",
            ...(text ? { content: text } : { content: null }),
            ...(toolCalls.length ? { tool_calls: toolCalls } : {}),
          };
        }
        return { role: "tool", tool_call_id: m.toolUseId, content: m.content };
      }),
    ];

    try {
      const stream = await client.chat.completions.create(
        {
          model,
          messages,
          ...(toolDefs.length ? { tools: toolDefs as unknown as OpenAI.ChatCompletionTool[] } : {}),
          stream: true,
          stream_options: { include_usage: true },
        },
        { signal: params.signal },
      );
      const toolBuffers = new Map<number, { id: string; name: string; args: string }>();
      let inputTokens = 0;
      let outputTokens = 0;
      let stopReason: "end_turn" | "tool_use" | "max_tokens" = "end_turn";
      for await (const chunk of stream) {
        if (params.signal?.aborted) return;
        const delta = chunk.choices[0]?.delta;
        if (!delta) continue;
        if (delta.content) yield { type: "text_delta", text: delta.content };
        for (const call of delta.tool_calls ?? []) {
          const idx = call.index;
          let buf = toolBuffers.get(idx);
          if (!buf) {
            buf = { id: call.id ?? `oa-${idx}`, name: call.function?.name ?? "", args: "" };
            toolBuffers.set(idx, buf);
            yield { type: "tool_use_start", id: buf.id, name: buf.name };
          }
          if (call.function?.arguments) {
            buf.args += call.function.arguments;
            yield {
              type: "tool_use_delta",
              id: buf.id,
              argsPartialJson: call.function.arguments,
            };
          }
        }
        const finish = chunk.choices[0]?.finish_reason;
        if (finish === "tool_calls") {
          for (const [, buf] of toolBuffers) yield { type: "tool_use_stop", id: buf.id };
          stopReason = "tool_use";
        }
        if (finish === "length") {
          stopReason = "max_tokens";
        }
        if (chunk.usage) {
          inputTokens = chunk.usage.prompt_tokens ?? inputTokens;
          outputTokens = chunk.usage.completion_tokens ?? outputTokens;
        }
      }
      yield { type: "usage", inputTokens, outputTokens };
      yield { type: "message_stop", stopReason };
    } catch (err) {
      if (params.signal?.aborted) return;
      yield { type: "error", error: err as Error };
    }
  }
}

File: src/docker_agent/services/api/providers/openai.py
"""OpenAI provider adapter.

Parity: ``src/services/api/providers/openai.ts``.
"""

from __future__ import annotations

import os
from collections.abc import AsyncIterator, Iterator
from typing import Any

from docker_agent.services.api._message_utils import build_openai_chat_messages
from docker_agent.services.api.tool_schema import to_openai_function
from docker_agent.services.api.types import (
    CallModelParams,
    ErrorEvent,
    MessageStopEvent,
    ProviderEvent,
    TextDeltaEvent,
    ToolUseDeltaEvent,
    ToolUseStartEvent,
    ToolUseStopEvent,
    UsageEvent,
)
from docker_agent.utils.sync_bridge import aiter_in_thread
from docker_agent.vault.api_key_store import ApiKeyStore, resolve_stored_api_key


class OpenAIProvider:
    name = "openai"

    def __init__(
        self,
        env: dict[str, str] | None = None,
        api_key_store: ApiKeyStore | None = None,
        *,
        client: Any | None = None,
    ) -> None:
        self._env = env or dict(os.environ)
        self._api_key_store = api_key_store
        self._client = client

    def _make_client(self, api_key: str) -> Any:
        if self._client is not None:
            return self._client
        from openai import OpenAI

        return OpenAI(
            api_key=api_key,
            base_url=self._env.get("OPENAI_BASE_URL") or None,
        )

    async def list_models(self) -> list[str]:
        import asyncio

        api_key = await resolve_stored_api_key("openai", self._env, self._api_key_store)
        if not api_key:
            raise RuntimeError("OPENAI_API_KEY not set")
        client = self._make_client(api_key)
        res = await asyncio.to_thread(client.models.list)
        return sorted(
            m.id for m in res.data if isinstance(getattr(m, "id", None), str) and m.id
        )

    async def stream(self, params: CallModelParams) -> AsyncIterator[ProviderEvent]:
        api_key = await resolve_stored_api_key("openai", self._env, self._api_key_store)
        if not api_key:
            yield ErrorEvent(error=RuntimeError("OPENAI_API_KEY not set"))
            return

        client = self._make_client(api_key)
        model = params.model or self._env.get("OPENAI_MODEL") or "gpt-4o-mini"
        tool_defs = [to_openai_function(t) for t in params.tools]
        messages = build_openai_chat_messages(params.messages, system=params.system)

        def sync_stream() -> Iterator[ProviderEvent]:
            try:
                stream = client.chat.completions.create(
                    model=model,
                    messages=messages,
                    tools=tool_defs or None,
                    stream=True,
                    stream_options={"include_usage": True},
                )
                tool_buffers: dict[int, dict[str, str]] = {}
                input_tokens = 0
                output_tokens = 0
                stop_reason: str = "end_turn"
                for chunk in stream:
                    delta = chunk.choices[0].delta if chunk.choices else None
                    if not delta:
                        continue
                    if delta.content:
                        yield TextDeltaEvent(text=delta.content)
                    for call in delta.tool_calls or []:
                        idx = call.index
                        buf = tool_buffers.get(idx)
                        if buf is None:
                            buf = {
                                "id": call.id or f"oa-{idx}",
                                "name": call.function.name or "",
                                "args": "",
                            }
                            tool_buffers[idx] = buf
                            yield ToolUseStartEvent(id=buf["id"], name=buf["name"])
                        if call.function.arguments:
                            buf["args"] += call.function.arguments
                            yield ToolUseDeltaEvent(
                                id=buf["id"], args_partial_json=call.function.arguments
                            )
                    finish = chunk.choices[0].finish_reason if chunk.choices else None
                    if finish == "tool_calls":
                        for buf in tool_buffers.values():
                            yield ToolUseStopEvent(id=buf["id"])
                        stop_reason = "tool_use"
                    if finish == "length":
                        stop_reason = "max_tokens"
                    if chunk.usage:
                        input_tokens = chunk.usage.prompt_tokens or input_tokens
                        output_tokens = chunk.usage.completion_tokens or output_tokens

                yield UsageEvent(input_tokens=input_tokens, output_tokens=output_tokens)
                yield MessageStopEvent(stop_reason=stop_reason)
            except Exception as err:  # noqa: BLE001 - surfaced as an event
                yield ErrorEvent(error=err)

        async for event in aiter_in_thread(sync_stream):
            yield event